Pocket/pocket-ios
# Settings > Premium Subscription - Show manage Subscription sheet

## Summary
- This PR adds the "Manage Subscription" sheet to Settings > Premium Subscription
- It also follows up on some open points discussed in #480
    - Inject a service that in turn makes the remote call for info
    - throws an error and displays an alert if the service request fails
    - uses a type to identify the subscription provider (source)

## Implementation Details
* Add `.manageSubscriptionsSheet` to `PremiumStatusView` to show the manage subscription sheet, for Apple subscribers
* Add `SubscriptionInfoService` and concrete implementation `PocketSubscriptionInfoService`
* Remove unused properties from SubscriptionInfo
* Tests: add PremiumStatusViewModelTests
* Tests: add MockSubscriptionInfoService

## Test Steps

> **Note**
check out the animations below for more insights on the testing steps

### Test skeleton
- checkout this branch then build/run
- login with a premium account
- Go to Settings then tap Premium Subscription
- Make sure that the skeleton in the subscription info appears before the actual info gets populated

### Test error thrown while retrieving subscription info
- checkout this branch
- to force an error, head [here](https://github.com/Pocket/pocket-ios/blob/796b6cb38cd32951d527677662169823c9d5ea9b/PocketKit/Sources/PocketKit/Settings/PremiumStatus/Model/SubscriptionInfoService.swift#L28) [`PocketSubscriptionService.getInfo()`] and inside `getInfo()`, throw an error instead of (or before) the call to `client.premiumStatus()`. This will force the call to fail and show the alert
- build/run and login with a premium account
- Go to Settings then tap Premium Subscription
- Make sure the error shown in the error animation appears
- Make sure that, when you tap "OK", the info screen dismisses

### Test manage subscription
> **Warning**
These testing steps are a bit tricky because it assumes to use test subscriptions on Apple via `StoreKit` configuration file, but if you have an active Apple subscription, you can skip steps 2, 3 and 4, provided that you are signed in with your Apple ID on the device of testing. Beware that if you make any changes to the subscription, though, these will be reflected on the actual Apple subscription.

1. checkout this branch
2. configure the `StoreKit` testing environment as described in #421 
3. tweak the property `isApple` [here](https://github.com/Pocket/pocket-ios/blob/796b6cb38cd32951d527677662169823c9d5ea9b/PocketKit/Sources/PocketKit/Settings/PremiumStatus/ViewModel/PremiumStatusViewModel.swift#L76), so that, regardless 
4. Head [here](https://github.com/Pocket/pocket-ios/blob/796b6cb38cd32951d527677662169823c9d5ea9b/PocketKit/Sources/PocketKit/PremiumUpgrade/View/PremiumUpgradeView.swift#L7) and in `PremiumUpgradeView`, set `shouldAllowUpgrade` to true: this will allow to make a test purchase.
5. Build/run and login with a free account, then head to Settings > Go Premium and purchase a subscription
6. Logout and login with a premium account from any source
7. Go to Settings > Premium Subscriptions
8. Tap Manage Subscriptions and make sure the Apple UI for managing subscriptions appears and it's actionable (see above warning in case you're using an actual Apple subscription)

## PR Checklist:

## Animations
Manage subscription - iPhone dark | Fetch subscription Info fails - iPad light
<img width=260 src=https://user-images.githubusercontent.com/34376330/224810081-7c5d099e-d3fc-4dfc-aecc-cd3cdca6b284.gif> | <img width=400 src=https://user-images.githubusercontent.com/34376330/224810126-34511a4e-736b-4d25-b453-6e085881f250.gif>
